src/test/java/org/opensearch/security/transport/SecurityInterceptorTests.java

```diff
@@ -201,17 +292,306 @@ public <T extends TransportResponse> void sendRequest(
+                TransportRequestOptions options,
+                TransportResponseHandler<T> handler
+            ) {
+                String serializedUserHeader = threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER);
+                assertEquals(serializedUserHeader, Base64Helper.serializeObject(user, useJDKSerialization));
+            }
+        };
+
+        securityInterceptor.sendRequestDecorate(finalSender, connection1, action, request, options, handler, localNode);
+
+        User transientUser4 = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        assertEquals(transientUser4, user);
+        assertEquals(threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER), null);
+
+        // from thread context inside sendRequestDecorate for local-node1 connection1
+        // this case is just for action trace logic validation
+        sender = new AsyncSender() {
+            @Override
+            public <T extends TransportResponse> void sendRequest(
+                Connection connection,
+                String action,
+                TransportRequest request,
+                TransportRequestOptions options,
+                TransportResponseHandler<T> handler
+            ) {
+                User transientUser = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+                assertEquals(transientUser, user);
+            }
+        };
+
+        // local node request
+        securityInterceptor.sendRequestDecorate(sender, connection1, action, request, options, handler, localNode);
+
+        // from original context
+        User transientUser5 = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        assertEquals(transientUser5, user);
+        assertEquals(threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER), null);
+
+        // even though we add the trace the restoring handler should remove it from the thread context
+        assertFalse(
+            threadPool.getThreadContext().getHeaders().keySet().stream().anyMatch(header -> header.startsWith("_opendistro_security_trace"))
+        );
     }
 
     /**
-     * Tests the scenario when remote node does not implement custom serialization protocol and uses JDK serialization
+     * This test populates all the thread context headers will fake values. This lets us verify the ensureCorrectHeaders
+     * method in the Security Interceptor.
      */
     @Test
-    public void testSendRequestDecorateWhenRemoteNodeUsesJDKSerde() {
-        testSendRequestDecorate(Version.V_2_0_0);
-    }
+    public void testSendRequestDecorateWithFakeHeaders() {
+        Version remoteNodeVersion = Version.CURRENT;
+        // Add a fake header to show it does not get misinterpreted
+        threadPool.getThreadContext().putHeader("FAKE_HEADER", "fake_value");
```

Comment:
For this test it seems like it would be fair easier to mock the thread pool and thread context, what do you think?

Reply:
I am not sure if that will trigger the coverage in the same way. In a couple places the interceptor checks against specific headers in a big conditional. In order to get full coverage it wants the negative and positive for each of these which is why I was adding each one like this. 

I can try swapping over and keep the changes if it seems like that will result in the same coverage.